src/runtime/set.cpp

```diff
@@ -118,22 +118,29 @@ BoxedSet* makeNewSet(BoxedClass* cls, Box* container) {
     assert(isSubclass(cls, frozenset_cls) || isSubclass(cls, set_cls));
 
     BoxedSet* rtn = new (cls) BoxedSet();
-
-    if (container) {
-        if (PyAnySet_Check(container)) {
-            for (auto&& elt : ((BoxedSet*)container)->s) {
-                rtn->s.insert(incref(elt));
-            }
-        } else if (PyDict_CheckExact(container)) {
-            for (auto&& elt : ((BoxedDict*)container)->d) {
-                rtn->s.insert(incref(elt.first));
-            }
-        } else {
-            for (auto elt : container->pyElements()) {
-                _setAddStolen(rtn, elt);
+    try {
+        if (container) {
+            if (PyAnySet_Check(container)) {
+                for (auto&& elt : ((BoxedSet*)container)->s) {
+                    rtn->s.insert(incref(elt));
+                }
+            } else if (PyDict_CheckExact(container)) {
+                for (auto&& elt : ((BoxedDict*)container)->d) {
+                    rtn->s.insert(incref(elt.first));
+                }
+            } else {
+                for (auto elt : container->pyElements()) {
+                    AUTO_DECREF(elt);
```

Comment:
I think the protocol should usually be that a ref-stealing function steals the ref even if an exception is thrown.  Which I think means here that maybe we should make this change inside _setAddStolen itself?